dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java

```diff
@@ -138,4 +164,70 @@ public boolean hasUserAccess( TrackedEntityInstance original, TrackedEntityInsta
 
         return true;
     }
+
+    private List<String> getMergeableAttributes( TrackedEntityInstance original, TrackedEntityInstance duplicate )
+    {
+        Map<String, String> existingTeavs = original.getTrackedEntityAttributeValues().stream()
+            .collect( Collectors.toMap( teav -> teav.getAttribute().getUid(), TrackedEntityAttributeValue::getValue ) );
+
+        List<String> attributes = new ArrayList<>();
+
+        for ( TrackedEntityAttributeValue teav : duplicate.getTrackedEntityAttributeValues() )
+        {
+            String existingVal = existingTeavs.get( teav.getAttribute().getUid() );
+
+            if ( existingVal != null )
+            {
+                if ( !existingVal.equals( teav.getValue() ) )
+                {
+                    throw new PotentialDuplicateConflictException(
+                        "Potential Duplicate contains conflicting value and cannot be merged." );
+                }
+            }
+            else
+            {
+                attributes.add( teav.getAttribute().getUid() );
+            }
+        }
+
+        return attributes;
+    }
+
+    private List<String> getMergeableRelationships( TrackedEntityInstance original, TrackedEntityInstance duplicate )
+    {
+        Map<String, Relationship> existingRelationships = original.getRelationshipItems().stream()
+            .collect( Collectors.toMap( ri -> ri.getRelationship().getUid(), RelationshipItem::getRelationship ) );
+
+        List<String> relationships = new ArrayList<>();
+
+        for ( RelationshipItem ri : duplicate.getRelationshipItems() )
+        {
+            Relationship existing = existingRelationships.get( ri.getRelationship().getUid() );
```

Comment:
The relationship uid is unique for all relationships, so what you are doing in your first if ( existing != null ) is saying if both original and duplicate have the same relationship. Which is great, and all you need. If they share uid, its means its a relationship between original and duplicate, which is the only one we block for. That also means you dont need the second equals check(ri.getRelationship) and you dont need the else block (TrackedEntityInstance from = ..)

This was a much better way to do what I explained complicated yesteday 😅